branchandbound.js

```diff
@@ -0,0 +1,143 @@
+var utils = require('./utils')
+
+var maxTries = 1000000
+
+function calculateEffectiveValues (utxos, feeRate) {
+  return utxos.map(function (utxo) {
+    if (isNaN(utils.uintOrNaN(utxo.value))) {
+      return {
+        utxo: utxo,
+        effectiveValue: 0
+      }
+    }
+
+    var effectiveFee = utils.inputBytes(utxo) * feeRate
+    var effectiveValue = utxo.value - effectiveFee
+    return {
+      utxo: utxo,
+      effectiveValue: effectiveValue
+    }
+  })
+}
+
+module.exports = function branchAndBound (utxos, outputs, feeRate) {
+  if (!isFinite(utils.uintOrNaN(feeRate))) return {}
+
+  var costPerOutput = utils.outputBytes({}) * feeRate
+  var costPerInput = utils.inputBytes({}) * feeRate
+  var costOfChange = costPerInput + costPerOutput
+
+  var outAccum = utils.sumOrNaN(outputs)
+
+  var effectiveUtxos = calculateEffectiveValues(utxos, feeRate).filter(function (x) {
+    return x.effectiveValue > 0
+  }).sort(function (a, b) {
+    return b.effectiveValue - a.effectiveValue
+  })
+
+  var selected = search(effectiveUtxos, outAccum, costOfChange)
+  if (selected != null) {
+    var inputs = []
+
+    for (var i = 0; i < effectiveUtxos.length; i++) {
+      if (selected[i]) {
+        inputs.push(effectiveUtxos[i].utxo)
+      }
+    }
+
+    return utils.finalize(inputs, outputs, feeRate)
+  } else {
+    const fee = feeRate * utxos.reduce(function (a, x) {
```

Comment:
Nope, the C++ doesn's seem to do that either.

It is not done in the coinselection.cpp - https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/coinselection.cpp

And it's not done in wallet.cpp - here the nvalue is just sum of the outputs
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2437

nvalueselect is just nvalue here (nFeeRet is 0 in the first pass, and BnB is used only in the first pass)
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2503

Here the SelectCoins is called, which calls (through some layers) BnB with the same target
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2555